File: python/tests/test_bitset.py
"""
Test suite for Bitset functionality.
"""

import pytest
import sys
import os
import threading

sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))
import posix_shm_py as shm


class TestBitset:
    """Test Bitset functionality."""
    
    def setup_method(self):
        """Set up test fixtures."""
        self.shm_name = "/test_bitset"
        self.mem = shm.SharedMemory(self.shm_name, 1024 * 1024)
    
    def teardown_method(self):
        """Clean up after tests."""
        try:
            self.mem.unlink()
        except:
            pass
    
    def test_basic_operations(self):
        """Test basic bitset operations."""
        bits = shm.Bitset1024(self.mem, "basic_bits")
        
        # Initially all bits should be false
        assert bits.none()
        assert not bits.any()
        assert not bits.all()
        assert bits.count() == 0
        assert len(bits) == 1024
        
        # Set some bits
        bits.set(0)
        bits.set(10)
        bits.set(100)
        bits.set(1023)  # Last bit
        
        assert bits.any()
        assert not bits.none()
        assert not bits.all()
        assert bits.count() == 4
        
        # Test individual bits
        assert bits.test(0)
        assert bits.test(10)
        assert bits.test(100)
        assert bits.test(1023)
        assert not bits.test(1)
        assert not bits.test(500)
    
    def test_set_with_value(self):
        """Test setting bits with explicit value."""
        bits = shm.Bitset1024(self.mem, "value_bits")
        
        # Set to true
        bits.set(5, True)
        assert bits.test(5)
        
        # Set to false
        bits.set(5, False)
        assert not bits.test(5)
        
        # Set multiple
        for i in range(0, 100, 10):
            bits.set(i, True)
        
        assert bits.count() == 10
    
    def test_reset_operations(self):
        """Test reset (clear) operations."""
        bits = shm.Bitset1024(self.mem, "reset_bits")
        
        # Set some bits
        for i in range(10):
            bits.set(i)
        assert bits.count() == 10
        
        # Reset individual bits
        bits.reset(0)
        bits.reset(5)
        bits.reset(9)
        
        assert not bits.test(0)
        assert not bits.test(5)
        assert not bits.test(9)
        assert bits.count() == 7
        
        # Reset all
        bits.reset_all()
        assert bits.none()
        assert bits.count() == 0
    
    def test_flip_operations(self):
        """Test flip operations."""
        bits = shm.Bitset1024(self.mem, "flip_bits")
        
        # Flip individual bits
        bits.flip(10)
        assert bits.test(10)
        
        bits.flip(10)
        assert not bits.test(10)
        
        # Set pattern and flip all
        for i in range(0, 1024, 2):
            bits.set(i)  # Set even bits
        
        count_before = bits.count()
        bits.flip_all()
        
        # All even bits should be false, odd bits true
        for i in range(1024):
            if i % 2 == 0:
                assert not bits.test(i)
            else:
                assert bits.test(i)
        
        assert bits.count() == 1024 - count_before
    
    def test_set_reset_all(self):
        """Test set_all and reset_all operations."""
        bits = shm.Bitset1024(self.mem, "all_bits")
        
        # Set all bits
        bits.set_all()
        assert bits.all()
        assert bits.count() == 1024
        
        for i in range(0, 1024, 100):
            assert bits.test(i)
        
        # Reset all bits
        bits.reset_all()
        assert bits.none()
        assert bits.count() == 0
        
        for i in range(0, 1024, 100):
            assert not bits.test(i)
    
    def test_find_operations(self):
        """Test find_first and find_next operations."""
        bits = shm.Bitset1024(self.mem, "find_bits")
        
        # Empty bitset
        assert bits.find_first() == 1024  # Not found
        
        # Set some bits
        bits.set(10)
        bits.set(20)
        bits.set(30)
        bits.set(100)
        
        # Find first
        assert bits.find_first() == 10
        
        # Find next
        pos = bits.find_first()
        assert pos == 10
        
        pos = bits.find_next(pos)
        assert pos == 20
        
        pos = bits.find_next(pos)
        assert pos == 30
        
        pos = bits.find_next(pos)
        assert pos == 100
        
        pos = bits.find_next(pos)
        assert pos == 1024  # No more bits
    
    def test_indexing_interface(self):
        """Test __getitem__ and __setitem__ interface."""
        bits = shm.Bitset1024(self.mem, "index_bits")
        
        # Test __setitem__
        bits[42] = True
        bits[100] = True
        bits[500] = False
        
        # Test __getitem__
        assert bits[42] == True
        assert bits[100] == True
        assert bits[500] == False
        assert bits[999] == False
        
        # Test with boolean values
        bits[42] = False
        assert not bits[42]
    
    def test_repr(self):
        """Test string representation."""
        bits = shm.Bitset1024(self.mem, "repr_bits")
        
        bits.set(1)
        bits.set(2)
        bits.set(3)
        
        repr_str = repr(bits)
        assert "Bitset1024" in repr_str
        assert "3/1024" in repr_str  # Shows count/size
    
    def test_boundary_conditions(self):
        """Test boundary conditions."""
        bits = shm.Bitset1024(self.mem, "boundary_bits")
        
        # Test first bit
        bits.set(0)
        assert bits.test(0)
        assert bits.find_first() == 0
        
        # Test last bit
        bits.set(1023)
        assert bits.test(1023)
        
        # Test out of bounds (should be safe)
        bits.set(2000)  # Should do nothing
        bits.reset(2000)  # Should do nothing
        bits.flip(2000)  # Should do nothing
        assert not bits.test(2000)  # Should return false
    
    def test_concurrent_operations(self):
        """Test concurrent bit operations."""
        bits = shm.Bitset1024(self.mem, "concurrent_bits")
        
        def set_bits(start, end):
            for i in range(start, end):
                bits.set(i)
        
        def flip_bits(start, end):
            for i in range(start, end):
                bits.flip(i)
        
        threads = []
        
        # Set different ranges in parallel
        for i in range(4):
            start = i * 250
            end = start + 250
            t = threading.Thread(target=set_bits, args=(start, end))
            threads.append(t)
            t.start()
        
        for t in threads:
            t.join()
        
        # All 1000 bits should be set
        assert bits.count() == 1000
    
    def test_persistence(self):
        """Test that bitset persists across connections."""
        # Set pattern
        bits1 = shm.Bitset1024(self.mem, "persist_bits")
        for i in range(0, 1024, 3):
            bits1.set(i)
        
        count1 = bits1.count()
        del bits1
        
        # Re-attach and verify
        bits2 = shm.Bitset1024(self.mem, "persist_bits")
        assert bits2.count() == count1
        
        for i in range(1024):
            if i % 3 == 0:
                assert bits2.test(i)
            else:
                assert not bits2.test(i)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/test_bitset.cpp
#include <catch2/catch_test_macros.hpp>
#include "zeroipc.h"
#include "bitset.h"
#include <thread>
#include <vector>
#include <bitset>

TEST_CASE("zeroipc::bitset basic operations", "[zeroipc::bitset]") {
    const std::string shm_name = "/test_bitset_basic";
    shm_unlink(shm_name.c_str());
    zeroipc::memory shm(shm_name, 10 * 1024 * 1024);

    SECTION("Create and use bitset") {
        zeroipc::bitset<100> bits(shm, "test_bits");
        
        REQUIRE(bits.size() == 100);
        REQUIRE(bits.count() == 0);
        REQUIRE(bits.none());
        REQUIRE(!bits.any());
        REQUIRE(!bits.all());
    }

    SECTION("Set and test individual bits") {
        zeroipc::bitset<64> bits(shm, "individual_bits");
        
        bits.set(0);
        bits.set(10);
        bits.set(63);
        
        REQUIRE(bits.test(0));
        REQUIRE(bits.test(10));
        REQUIRE(bits.test(63));
        REQUIRE(!bits.test(1));
        REQUIRE(!bits.test(32));
        
        REQUIRE(bits.count() == 3);
        REQUIRE(bits.any());
        REQUIRE(!bits.none());
        REQUIRE(!bits.all());
    }

    SECTION("Reset bits") {
        zeroipc::bitset<32> bits(shm, "reset_bits");
        
        bits.set(5);
        bits.set(10);
        REQUIRE(bits.count() == 2);
        
        bits.reset(5);
        REQUIRE(!bits.test(5));
        REQUIRE(bits.test(10));
        REQUIRE(bits.count() == 1);
        
        bits.reset(10);
        REQUIRE(bits.none());
    }

    SECTION("Flip bits") {
        zeroipc::bitset<16> bits(shm, "flip_bits");
        
        bits.flip(3);
        REQUIRE(bits.test(3));
        REQUIRE(bits.count() == 1);
        
        bits.flip(3);
        REQUIRE(!bits.test(3));
        REQUIRE(bits.count() == 0);
    }

    SECTION("Set with value") {
        zeroipc::bitset<8> bits(shm, "set_value");
        
        bits.set(2, true);
        REQUIRE(bits.test(2));
        
        bits.set(2, false);
        REQUIRE(!bits.test(2));
    }

    SECTION("Set all bits") {
        zeroipc::bitset<100> bits(shm, "set_all");
        
        bits.set();
        REQUIRE(bits.count() == 100);
        REQUIRE(bits.all());
        REQUIRE(!bits.none());
        
        // Individual bits should all be set
        for (size_t i = 0; i < 100; ++i) {
            REQUIRE(bits.test(i));
        }
    }

    SECTION("Reset all bits") {
        zeroipc::bitset<50> bits(shm, "reset_all");
        
        // Set some bits first
        for (size_t i = 0; i < 50; i += 3) {
            bits.set(i);
        }
        REQUIRE(bits.any());
        
        bits.reset();
        REQUIRE(bits.count() == 0);
        REQUIRE(bits.none());
        REQUIRE(!bits.any());
    }

    SECTION("Flip all bits") {
        zeroipc::bitset<10> bits(shm, "flip_all");
        
        bits.set(0);
        bits.set(2);
        bits.set(4);
        REQUIRE(bits.count() == 3);
        
        bits.flip();
        REQUIRE(bits.count() == 7);
        REQUIRE(!bits.test(0));
        REQUIRE(bits.test(1));
        REQUIRE(!bits.test(2));
        REQUIRE(bits.test(3));
    }

    SECTION("Find first set bit") {
        zeroipc::bitset<128> bits(shm, "find_first");
        
        REQUIRE(bits.find_first() == 128);  // No bits set
        
        bits.set(50);
        bits.set(30);
        bits.set(70);
        
        REQUIRE(bits.find_first() == 30);
    }

    SECTION("Find next set bit") {
        zeroipc::bitset<100> bits(shm, "find_next");
        
        bits.set(10);
        bits.set(20);
        bits.set(30);
        bits.set(40);
        
        size_t pos = bits.find_first();
        REQUIRE(pos == 10);
        
        pos = bits.find_next(pos);
        REQUIRE(pos == 20);
        
        pos = bits.find_next(pos);
        REQUIRE(pos == 30);
        
        pos = bits.find_next(pos);
        REQUIRE(pos == 40);
        
        pos = bits.find_next(pos);
        REQUIRE(pos == 100);  // No more bits
    }

    shm.unlink();
}

TEST_CASE("zeroipc::bitset bitwise operations", "[zeroipc::bitset]") {
    const std::string shm_name = "/test_bitset_bitwise";
    shm_unlink(shm_name.c_str());
    zeroipc::memory shm(shm_name, 10 * 1024 * 1024);

    zeroipc::bitset<64> bits1(shm, "bits1");
    zeroipc::bitset<64> bits2(shm, "bits2");
    
    // Setup bit patterns
    for (size_t i = 0; i < 64; i += 2) {
        bits1.set(i);  // Even bits
    }
    for (size_t i = 0; i < 64; i += 3) {
        bits2.set(i);  // Every third bit
    }

    SECTION("Bitwise AND") {
        zeroipc::bitset<64> result(shm, "and_result");
        result.set();  // Start with all 1s
        
        result &= bits1;
        result &= bits2;
        
        // Should only have bits set where both bits1 and bits2 are set
        for (size_t i = 0; i < 64; ++i) {
            bool expected = (i % 2 == 0) && (i % 3 == 0);
            REQUIRE(result.test(i) == expected);
        }
    }

    SECTION("Bitwise OR") {
        zeroipc::bitset<64> result(shm, "or_result");
        
        result |= bits1;
        result |= bits2;
        
        // Should have bits set where either bits1 or bits2 are set
        for (size_t i = 0; i < 64; ++i) {
            bool expected = (i % 2 == 0) || (i % 3 == 0);
            REQUIRE(result.test(i) == expected);
        }
    }

    SECTION("Bitwise XOR") {
        zeroipc::bitset<64> result(shm, "xor_result");
        result |= bits1;  // Copy bits1
        
        result ^= bits2;
        
        // Should have bits set where bits1 XOR bits2
        for (size_t i = 0; i < 64; ++i) {
            bool expected = (i % 2 == 0) != (i % 3 == 0);
            REQUIRE(result.test(i) == expected);
        }
    }

    shm.unlink();
}

TEST_CASE("zeroipc::bitset concurrent operations", "[zeroipc::bitset][concurrent]") {
    const std::string shm_name = "/test_bitset_concurrent";
    shm_unlink(shm_name.c_str());
    zeroipc::memory shm(shm_name, 10 * 1024 * 1024);

    SECTION("Concurrent bit setting") {
        zeroipc::bitset<10000> bits(shm, "concurrent_set");
        const int num_threads = 4;
        const int bits_per_thread = 2500;
        
        std::vector<std::thread> threads;
        
        for (int t = 0; t < num_threads; ++t) {
            threads.emplace_back([&bits, t, bits_per_thread]() {
                for (int i = 0; i < bits_per_thread; ++i) {
                    bits.set(t * bits_per_thread + i);
                }
            });
        }
        
        for (auto& t : threads) {
            t.join();
        }
        
        REQUIRE(bits.count() == 10000);
        REQUIRE(bits.all());
    }

    SECTION("Concurrent flipping") {
        zeroipc::bitset<1000> bits(shm, "concurrent_flip");
        const int num_threads = 4;
        const int flips_per_thread = 1000;
        
        std::vector<std::thread> threads;
        
        for (int t = 0; t < num_threads; ++t) {
            threads.emplace_back([&bits, t, flips_per_thread]() {
                for (int i = 0; i < flips_per_thread; ++i) {
                    // Each thread flips its own range
                    size_t bit = (t * 250) + (i % 250);
                    bits.flip(bit);
                }
            });
        }
        
        for (auto& t : threads) {
            t.join();
        }
        
        // Each bit flipped 4 times (1000/250), so should be back to 0
        REQUIRE(bits.none());
    }

    SECTION("Concurrent read/write") {
        zeroipc::bitset<10000> bits(shm, "concurrent_rw");
        
        // Pre-set some bits
        for (size_t i = 0; i < 10000; i += 2) {
            bits.set(i);
        }
        
        std::atomic<int> read_count{0};
        std::vector<std::thread> threads;
        
        // Writers
        for (int t = 0; t < 2; ++t) {
            threads.emplace_back([&bits, t]() {
                for (int i = 0; i < 5000; ++i) {
                    bits.flip(t * 5000 + i);
                }
            });
        }
        
        // Readers
        for (int t = 0; t < 2; ++t) {
            threads.emplace_back([&bits, &read_count]() {
                for (int i = 0; i < 10000; ++i) {
                    for (size_t j = 0; j < 10000; ++j) {
                        if (bits.test(j)) {
                            read_count++;
                        }
                    }
                }
            });
        }
        
        for (auto& t : threads) {
            t.join();
        }
        
        // Readers should have seen some bits set
        REQUIRE(read_count > 0);
    }

    shm.unlink();
}

TEST_CASE("zeroipc::bitset edge cases", "[zeroipc::bitset]") {
    const std::string shm_name = "/test_bitset_edge";
    shm_unlink(shm_name.c_str());
    zeroipc::memory shm(shm_name, 10 * 1024 * 1024);

    SECTION("Non-64-bit aligned size") {
        zeroipc::bitset<73> bits(shm, "odd_size");
        
        // Should handle bits near boundary correctly
        bits.set(63);
        bits.set(64);
        bits.set(72);
        
        REQUIRE(bits.test(63));
        REQUIRE(bits.test(64));
        REQUIRE(bits.test(72));
        
        // Out of bounds should be safe
        bits.set(100);  // Should do nothing
        REQUIRE(!bits.test(73));
        
        // Set all should respect size
        bits.set();
        REQUIRE(bits.count() == 73);
    }

    SECTION("Single bit") {
        zeroipc::bitset<1> single(shm, "single_bit");
        
        REQUIRE(!single.test(0));
        single.set(0);
        REQUIRE(single.test(0));
        REQUIRE(single.all());
        REQUIRE(single.any());
        REQUIRE(!single.none());
        
        single.flip();
        REQUIRE(single.none());
    }

    SECTION("Large bitset") {
        zeroipc::bitset<100000> large(shm, "large_bits");
        
        // Set pattern
        for (size_t i = 0; i < 100000; i += 1000) {
            large.set(i);
        }
        
        REQUIRE(large.count() == 100);
        
        // Find operations should work
        REQUIRE(large.find_first() == 0);
        REQUIRE(large.find_next(0) == 1000);
    }

    shm.unlink();
}

TEST_CASE("zeroipc::bitset cross-process", "[zeroipc::bitset][process]") {
    const std::string shm_name = "/test_bitset_process";
    shm_unlink(shm_name.c_str());
    
    SECTION("Bitset persistence across processes") {
        // Process 1: Create and set bits
        {
            zeroipc::memory shm1(shm_name, 1024 * 1024);
            zeroipc::bitset<256> bits(shm1, "persistent_bits");
            
            for (size_t i = 0; i < 256; i += 4) {
                bits.set(i);
            }
            
            REQUIRE(bits.count() == 64);
        }
        
        // Process 2: Open and verify
        {
            zeroipc::memory shm2(shm_name, 0);  // Attach only
            zeroipc::bitset<256> bits(shm2, "persistent_bits");
            
            REQUIRE(bits.count() == 64);
            
            for (size_t i = 0; i < 256; ++i) {
                if (i % 4 == 0) {
                    REQUIRE(bits.test(i));
                } else {
                    REQUIRE(!bits.test(i));
                }
            }
            
            // Modify
            bits.flip();
        }
        
        // Process 3: Verify modifications
        {
            zeroipc::memory shm3(shm_name, 0);
            zeroipc::bitset<256> bits(shm3, "persistent_bits");
            
            REQUIRE(bits.count() == 192);  // 256 - 64
            
            for (size_t i = 0; i < 256; ++i) {
                if (i % 4 == 0) {
                    REQUIRE(!bits.test(i));  // These were flipped
                } else {
                    REQUIRE(bits.test(i));   // These are now set
                }
            }
        }
    }
    
    shm_unlink(shm_name.c_str());
}